service/storage_service.cc

```diff
@@ -1875,7 +1875,7 @@ future<> storage_service::join_token_ring(cdc::generation_service& cdc_gen_servi
             slogger.info("Replacing a node with {} IP address, my address={}, node being replaced={}",
                 get_broadcast_address() == *replace_address ? "the same" : "a different",
                 get_broadcast_address(), *replace_address);
-            tmptr->update_topology(*replace_address, std::move(ri->dc_rack), locator::node::state::leaving);
+            tmptr->update_topology(*replace_address, std::move(ri->dc_rack), locator::node::state::being_replaced);
```

Comment:
If a node replaces using the same ip address of the node being replaced, should we call it being_replaced or replacing?

Reply:
If we mark it as being_replaced,  is_token_owner_for_read() is going to return true for the node being replaced which is supposed to be dead and we should not read from it from the perspective node ops. 

```
    bool is_token_owner_for_read() const noexcept {
        switch (_state) {
        case state::normal:
        case state::being_decommissioned:
        case state::being_replaced:
            return true;
        default:
            return false;
        }
    }
```

Reply:
> If a node replaces using the same ip address of the node being replaced, should we call it being_replaced or replacing?

Now that we don't reuse the replaced node host id, there will be 2 node objects in topology, distinguished by their host_id.
The one which state is updated here is always the one being replaced.

Reply:
> If we mark it as being_replaced, is_token_owner_for_read() is going to return true for the node being replaced which is supposed to be dead and we should not read from it from the perspective node ops.

You're 100% right. It's a bug and it shouldn't be returned there.

Reply:
fixed in v3

Reply:
On one hand, removing being_replaced from is_token_owner() will exclude the node from nodes to sync in replace. On the other hand, node being_replaced is still the owner of the tokens, it is wrong not to include it. 

I think it is better we keep the token ownership correct and filter the node being replaced with the filter. 

Actually we already have such filter

    // Step 1: Decide who needs to sync data for replace operation
    // The replacing node is not a normal token owner yet
    // Add it back explicitly after checking all other nodes.
    ctl.start("replace", [&] (gms::inet_address node) {
        return node != replace_address;
    });

Reply:
But isn't the replaced node required to be dead to be replaced?  If so, it shouldn't be sync'ed with.
Also, if we include replaced nodes we should include also removed nodes since they are officially still token owners until removed / replaced.  The point of `is_token_owner_for_read` was to restrict to nodes that should be accessed for read - and replaced and removed nodes are not part of that set.

Reply:
Note: replaced may be in the future, with tablets, if we allow replacing live nodes.
But since bootstrapping a new node with tablets is quick and adds the new node as empty - and only later uses the tablets load balancer to fill it, it will be feasible to add one or more new nodes and then simply decommission the node we want to replace.  That would be a close enough approximation of replacing a live node.
/Cc @tgrabiec